File: LC206-ReverseLinkedList.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode* prev = NULL;
    struct ListNode* nxt = NULL;
    struct ListNode* curr = head;

    while (curr) {
        nxt = curr->next;
        curr->next = prev;
        prev = curr;
        curr = nxt;
    }
    return prev;
}


// Solution with 3 variable (prev, nxt, curr) and 4 steps per iteration in while loop.
// Time complexity: O(n) - We traverse the entire list once.
// Space complexity: O(1) - We use a constant amount of extra space.

File: LC206-ReverseLinkedList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head

        while curr:
            nxt = curr.next      # Save the next node
            curr.next = prev     # Reverse the link
            prev = curr          # Move prev to curr
            curr = nxt           # Move to the next node

        # prev now points to the new head of the reversed list
        return prev
